python/cuspatial/cuspatial/geometry/geodataframe.py

```diff
@@ -0,0 +1,100 @@
+
+import cudf
+
+from cuspatial.geometry.geoarrowbuffers import GeoArrowBuffers
+from cuspatial.geometry.geocolumn import GeoColumn, GeoPandasMeta
+from cuspatial.geometry.geoseries import GeoSeries
+from cuspatial.geometry.geoutil import is_geometry_type
+from cuspatial.io.geopandas_adapter import GeoPandasAdapter
+
+
+class GeoDataFrame(cudf.DataFrame):
+    """
+    A GPU GeoDataFrame object.
+    """
+
+    def __init__(self, data: gpGeoDataFrame = None):
+        """
+        Constructs a GPU GeoDataFrame from a GeoPandas dataframe.
+
+        Parameters
+        ----------
+        data : A geopandas.geodataframe.GeoDataFrame object
+        """
+        super().__init__()
+        if isinstance(data, gpGeoDataFrame):
+            self.index = data.index
+            for col in data.columns:
+                if is_geometry_type(data[col]):
+                    adapter = GeoPandasAdapter(data[col])
+                    buffers = GeoArrowBuffers(
+                        adapter.get_geoarrow_host_buffers()
+                    )
+                    pandas_meta = GeoPandasMeta(adapter.get_geopandas_meta())
+                    column = GeoColumn(buffers, pandas_meta)
+                    self._data[col] = column
+                else:
+                    self[col] = data[col]
+        elif data is None:
+            pass
+        else:
+            raise ValueError("Invalid type passed to GeoDataFrame ctor")
+
+    @property
+    def _constructor(self):
+        return GeoDataFrame
+
+    @property
+    def _constructor_sliced(self):
+        return _GeoSeriesUtility
+
+    def to_pandas(self, nullable=False):
+        """
+        Calls `self.to_geopandas`, converting GeoSeries columns into GeoPandas
+        columns and cudf.Series columns into pandas.Series columns, and
+        returning a pandas.DataFrame.
+
+        Parameters
+        ----------
+        nullable: matches the cudf `to_pandas` signature, not yet supported.
+        """
+        return self.to_geopandas(nullable=nullable)
+
+    def to_geopandas(self, nullable=False):
+        """
+        Returns a new GeoPandas GeoDataFrame object from the coordinates in
+        the cuspatial GeoDataFrame.
+
+        Parameters
+        ----------
+        nullable: matches the cudf `to_pandas` signature, not yet supported.
+        """
+        if nullable is True:
+            raise ValueError("cuGeoDataFrame doesn't support N/A yet")
+        result = gpGeoDataFrame(
+            dict([(col, self[col].to_pandas()) for col in self.columns]),
+            index=self.index.to_pandas(),
+        )
+        return result
+
+    def __repr__(self):
+        return self.to_pandas().__repr__() + "\n" + "(GPU)" + "\n"
+
+    def groupby(self, *args, **kwargs):
+        result = super().groupby(*args, **kwargs)
+        for col in self.columns:
+            if is_geometry_type(self[col]):
+                result.obj = result.obj.drop(col, axis=1)
+        return result
```

Comment:
I hate to even suggest this but I imagine this pattern will proliferate where maybe we do something like a `getattr` hack where we can handle this for any function?

Reply:
If you are suggesting that we override `getattr`, then create a list of function names that drop geometry columns when called in pandas, doing so whenever there's a match, then I suggest we postpone that until a later PR. :)

Reply:
I guess I'm suggesting we always override `getattr` and whenever the returned object from a `getattr` is a `DataFrame`, `Series`, or `Index`, we postprocess it to get the geometry types back, but this is a problem more likely solved in cuDF as opposed to here. Agreed on punting.

Reply:
I think I'm already doing what you want with the inheritance from `NumericColumn` and the use of `_copy_type_metadata` - Any existing CUDF algorithm just treats the geometry column as a `Numeric`, sorts or rearranges it appropriately, then `GeoColumn` treats that as a hidden index to reorder the geometry.

I suppose in the case of `groupby`, the geometry column is dropped because most columnar reductions don't make sense with geometry, like `min/max`, etc.

Reply:
Agree to punt.